File: src/Systems/Swift/facade.swift
struct AuthService {
    func authenticate(_ user: String) -> String { "auth(\(user))" }
}

struct InventoryService {
    func reserve(_ sku: String) -> String { "reserve(\(sku))" }
}

struct BillingService {
    func charge(_ cents: Int) -> String { "charge(\(cents))" }
}

struct CheckoutFacade {
    let auth: AuthService
    let inventory: InventoryService
    let billing: BillingService

    func checkout(user: String, sku: String, cents: Int) -> String {
        "checkout=\(auth.authenticate(user))>\(inventory.reserve(sku))>\(billing.charge(cents))"
    }
}

let facade = CheckoutFacade(auth: AuthService(), inventory: InventoryService(), billing: BillingService())
print(facade.checkout(user: "alice", sku: "SKU-42", cents: 499))

File: src/Systems/Rust/facade.rs
struct AuthService;
impl AuthService {
    fn authenticate(&self, user: &str) -> String { format!("auth({user})") }
}

struct InventoryService;
impl InventoryService {
    fn reserve(&self, sku: &str) -> String { format!("reserve({sku})") }
}

struct BillingService;
impl BillingService {
    fn charge(&self, cents: i32) -> String { format!("charge({cents})") }
}

struct CheckoutFacade {
    auth: AuthService,
    inventory: InventoryService,
    billing: BillingService,
}

impl CheckoutFacade {
    fn checkout(&self, user: &str, sku: &str, cents: i32) -> String {
        [
            self.auth.authenticate(user),
            self.inventory.reserve(sku),
            self.billing.charge(cents),
        ]
        .join(">")
    }
}

fn main() {
    let facade = CheckoutFacade {
        auth: AuthService,
        inventory: InventoryService,
        billing: BillingService,
    };
    println!("checkout={}", facade.checkout("alice", "SKU-42", 499));
}
